pythonPackages.sanic: 19.3.1 -> 19.6.3 (and fix websockets and other issues)

simonchatts opened the pull request:
###### Motivation for this change

In `pythonPackages`, the `sanic` package requires `websockets` version 6.x, but the default provided is version 7.0.

The current workaround is just to [delete the version bounds](https://github.com/NixOS/nixpkgs/blob/1a1a7ed26652545dc6a2e2bd02f7812e1b37d196/pkgs/development/python-modules/sanic/default.nix#L51) on this dependency. This moves the problem to runtime: the first time the websocket functionality is actually used, it fails (see below for a test case and resulting crash).

The change here brings in the correct version of the `websockets` Python package, so that this works.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Test case

In case useful, here's a minimal concrete test case.

If this is `server.py`:
```python
from sanic import Sanic
from sanic.websocket import WebSocketProtocol

app = Sanic()

@app.websocket('/ws')
async def foo(req, ws):
  await ws.send('hello from websocket!')

app.run(protocol=WebSocketProtocol)
```
and you run this server via
```
nix-shell --pure -p 'python3.withPackages(p: [p.sanic])' --command 'python server.py'
```
(along with an assumed `-I nixpkgs=<path to nixpkgs>`), and invoke a test client:
```
nix-shell -p websocat --command 'websocat ws://127.0.0.1:8000/ws'
```
then with the current code, you get a crashdump:
```
[2019-09-08 19:35:50 +0100] [1688] [ERROR] Exception occurred while handling uri: 'ws://127.0.0.1:8000/ws'
Traceback (most recent call last):
  File "/nix/store/034m3rj9j27zvdldvpn6bdzpm0h3fn7c-python3-3.7.4-env/lib/python3.7/site-packages/sanic/app.py", line 917, in handle_request
    response = await response
  File "/nix/store/034m3rj9j27zvdldvpn6bdzpm0h3fn7c-python3-3.7.4-env/lib/python3.7/site-packages/sanic/app.py", line 473, in websocket_handler
    ws = await protocol.websocket_handshake(request, subprotocols)
  File "/nix/store/034m3rj9j27zvdldvpn6bdzpm0h3fn7c-python3-3.7.4-env/lib/python3.7/site-packages/sanic/websocket.py", line 69, in websocket_handshake
    key = handshake.check_request(request.headers)
  File "/nix/store/034m3rj9j27zvdldvpn6bdzpm0h3fn7c-python3-3.7.4-env/lib/python3.7/site-packages/websockets/handshake.py", line 82, in check_request
    [parse_connection(value) for value in headers.get_all('Connection')], []
AttributeError: 'multidict._multidict.CIMultiDict' object has no attribute 'get_all'
```
With the code changes here, you just see the correct client output:
```
hello from websocket!
```

###### Notify maintainers

cc @costrouc

simonchatts committed:
pythonPackages.sanic: 19.3.1 -> 19.6.3

The new version depend on two packages that are not currently in nixpkgs, and
marked as already deprecated by their author. So just include them locally. Also,
adapt to the fact that the version of websockets in nixpkgs is 8.x, when sanic
wants 7.x.

FRidh reviewed (changes requested):
Multiple versions of a Python package in the Python packages set is not allowed.

FRidh commented:
Maybe bumping the version of sanic is a solution.

simonchatts commented:
Ah yes, makes sense now you say it. The very latest release of sanic updates to the version of websockets in nixpkg - I’ll give it a test, and update this PR if it seems stable. Thanks for the feedback!

simonchatts committed:
pythonPackages.uvicorn: fix propagated LICENSE.md

simonchatts commented:
OK, I've updated sanic to the latest release. This required several other cascading changes:

 - The newer version of `sanic` now depends on `requests-async`, which in turn depends on `httpcore`. These aren't in nixpkgs, but they are also marked as deprecated by their author. So rather than introduce an already-dying maintenance burden, I've declared them just as local packages within `sanic`.
 - The `h11` Python package in the dependency graph was already broken for Python 3, so fixed that (by applying a one-line patch from their repo that has yet to be released).
 - The `uvicorn` framework is used for `sanic` testing, and this was pinned to an old version of `uvloop`. I've updated this, and tested with all the other packages that depend on `uvicorn`.

I've tested the result on both nixos and macOS, and everything looks good so far.

lheckemann commented:
Is this a fix that should be backported to 19.09?

simonchatts commented:
@lheckemann I assume this is tradeoff judgment aimed at the reviewers and code owners. But I can say that without some sort of changes like these, in 19.09:

 - `h11` is broken for python3 (and hence dependent packages like `uvicorn`, `fastapi`, `starlette` and `wsproto`)
 - `sanic`’s websockets functionality is broken

wd15 commented:
The datasette tests don't seem to be passing with this change. However, I updated datasette to the lastest version and it builds and passes the tests (sanic is no longer a dependency). [Here are the changes](https://github.com/wd15/nixpkgs/commit/9916b420124d0d9d20a0d2e8a295a3aa622f571d). Shall I try and add that into this PR or feel free to just duplicate?

simonchatts committed:
pythonPackages.datasette: 0.28 -> 0.29.3

Patch donated by @wd51, and removes sanic dependency.

simonchatts commented:
Thanks @wd15! I've just duplicated as suggested to keep things simple.

I've also checked that the other python package that depends on sanic (`pytest-sanic`, in addition to `datasette`) at least builds with these changes.

wd15 commented:
After merging in the latest master (2ff3620d0dcd) into this branch and resolving h11 conflicts nix-review passes so I'm giving my approval to this PR.

Result of `nix-review` [1](https://github.com/Mic92/nix-review) using `Linux 5.0.0-27-generic #28-Ubuntu x86_64 GNU/Linux`
<details>
  <summary>5 packages were built:</summary>

  - datasette (python37Packages.datasette)
  - python37Packages.fastapi (python37Packages.datasette)
  - python37Packages.sanic (python37Packages.datasette)
  - python37Packages.starlette (python37Packages.datasette)
  - python37Packages.uvicorn (python37Packages.datasette)
</details>

zimbatm commented:
it's rare to see so detailed reports in a PR. <3 @simonchatts 

It looks like h11 has been fixed for python3 in master already. Do you mind rebasing the PR one last time?

simonchatts commented:
Thanks @zimbatm - rebased with the existing h11 fix.

zimbatm commented on pkgs/development/python-modules/uvicorn/default.nix:
```diff
@@ -35,6 +35,11 @@ buildPythonPackage rec {
+  # LICENCE.md gets propagated without this, causing collisions
+  postInstall = ''
+    rm $out/LICENSE.md
+  '';
```
@FRidh do you know why the LICENSE.md files get copied to the top-level like that?

wd15 replied:
It seems to be [listed](https://github.com/encode/uvicorn/blob/141a817d66bfa3c5a3d0810fd2fa6a239f5f187e/setup.py#L64) in Uvicorn's `setup.py` as a `data_file` so [this happens](https://docs.python.org/2/distutils/setupscript.html#installing-additional-files). Somebody has already filed an [issue](https://github.com/encode/uvicorn/issues/392).

zimbatm replied:
Thanks for the explanation. It's curious that `requests-async` and `httpcore` also suffer from the same issue (both in this PR as well). But as long as it's not a widespread issue I guess it's ok to leave the postInstall here.

Mic92 commented:
looks there is a conflict again.

zimbatm reviewed (commented):
one last rebase?

zimbatm commented on pkgs/development/python-modules/uvicorn/default.nix:
```diff
@@ -35,6 +35,11 @@ buildPythonPackage rec {
 
   patches = [ ./setup.patch ];
 
+  # LICENCE.md gets propagated without this, causing collisions
```
```suggestion
  # LICENCE.md gets propagated without this, causing collisions
  # see https://github.com/encode/uvicorn/issues/392
```

simonchatts replied:
Added

simonchatts commented:
Thanks @zimbatm! I've done the rebase and suggested comment in my local repo, but hit a problem: @jonringer [recently upgraded](https://github.com/NixOS/nixpkgs/commit/9b092e228b2d9eab54c49685cddd6e30fde8eceb) the Python websockets from 7.x to 8.x. Sanic [requires 7.x](https://github.com/huge-success/sanic/blob/21ebf6d7772365c6e621a1c2a0729a2a9df86e8a/environment.yml#L11).

I presume this sort of issue has come up before - any advice @wd15 @FRidh?

BTW @jonringer that websockets 8.x upgrade also breaks on Darwin, since it also enables the tests but they don't work on Darwin (hit `OSError: AF_UNIX path too long`) but that's a minor detail here.

jonringer commented:
oops :( I got added to extra-users recently to ofborg,  so I'll be more aware of when darwin fails

simonchatts committed:
pythonPackages.websocket: disable failing tests on darwin

simonchatts commented:
I think I have a resolution to the major version conflict for `websockets`: the [changes for websockets 8.x](https://websockets.readthedocs.io/en/stable/changelog.html) are actually very tame for sanic's usage, and they are already almost compatible.

So I've updated this branch with a small modification to sanic to use the 8.x websockets now in nixpkgs. (I'll also try to upstream these, so this should all be simpler in the future.) I've tested these with some apps that make heavy use of sanic websockets, and it all looks fine so far.

@jonringer I've also fixed the Darwin test issue with websockets here, but can obviously drop this if it's already covered by some other commit.

I think this resolves all the pending issues.

Mic92 commented:
```
@GrahamcOfBorg build python37Packages.datasette datasette python37Packages.sanic python37Packages.uvicorn
```
